File: lib/data/vault.dart
import 'dart:convert';

import 'package:canarslan_website/services/vault_crypto_stub.dart'
    if (dart.library.js_interop) 'package:canarslan_website/services/vault_crypto_web.dart';
import 'package:flutter/foundation.dart';
import 'package:http/http.dart' as http;

/// One page behind the passcode.
@immutable
class VaultPage {
  const VaultPage({required this.title, required this.body});

  final String title;
  final List<String> body;
}

/// The private pages, and the one way in.
///
/// Everything here is deliberately content rather than code. The pages live in
/// `web/vault.json` as a single AES-GCM blob, which means the public repository
/// holds no trace of them and neither does the compiled bundle — a reader who
/// downloads either gets ciphertext. Adding a page is editing a local file and
/// re-running `tool/vault.js`; it never touches Dart, so it never touches a
/// commit.
///
/// **What this does not do.** The blob ships to the browser, so anyone can take
/// it away and try passcodes against it offline, as fast as their hardware
/// allows. A high PBKDF2 count makes each attempt expensive, but six digits is
/// a million attempts and that is a short afternoon for a GPU. This keeps the
/// pages out of the repository and away from anyone reading the bundle. It is
/// not a place for anything that would actually hurt to lose.
abstract class Vault {
  /// Where the encrypted file sits once the build has copied `web/` across.
  /// Absent on a checkout that has never run the tool, which is the normal
  /// state of the public repository.
  static const _file = 'vault.json';

  static const _timeout = Duration(seconds: 8);

  static List<VaultPage>? _open;

  /// The pages, once unlocked. Null until then.
  ///
  /// Kept in memory only. Nothing decrypted is written to storage and the
  /// passcode is never stored at all, so closing the tab locks the door again.
  static List<VaultPage>? get pages => _open;

  static bool get isOpen => _open != null;

  static void close() => _open = null;

  /// Fetches the blob and tries [passcode] against it.
  ///
  /// Returns false for a wrong code, a missing file and a corrupt one alike.
  /// The gate cannot tell them apart, and telling a visitor which one it was
  /// would only confirm the vault exists.
  static Future<bool> unlock(String passcode) async {
    try {
      final response = await http
          .get(Uri.base.resolve(_file))
          .timeout(_timeout);
      if (response.statusCode != 200) return false;

      final envelope = jsonDecode(response.body) as Map<String, dynamic>;
      final plain = await decryptVaultImpl(
        passcode: passcode,
        salt: _bytes(envelope['salt']),
        iv: _bytes(envelope['iv']),
        payload: _bytes(envelope['data']),
        iterations: (envelope['iterations'] as num).toInt(),
      );
      if (plain == null) return false;

      final decoded = jsonDecode(plain) as Map<String, dynamic>;
      _open = [
        for (final page in decoded['pages'] as List? ?? const [])
          VaultPage(
            title: (page as Map)['title']?.toString() ?? '',
            body: [
              for (final line in page['body'] as List? ?? const [])
                line.toString(),
            ],
          ),
      ];
      return true;
    } catch (_) {
      return false;
    }
  }

  static Uint8List _bytes(Object? base64Value) =>
      base64Decode(base64Value! as String);
}

File: tool/vault.js
#!/usr/bin/env node
// Encrypts the private pages into web/vault.json.
//
//   node tool/vault.js private/pages.json <passcode>
//   node tool/vault.js private/a.json <code> private/b.json <other code>
//
// Arguments come in pairs: a document, and the passcode that opens it. Each
// pair becomes a compartment, and the gate opens whichever one the code
// belongs to. **Every pair has to be on the one command line** — the file is
// written whole each time, so a run that names one document leaves the site
// with one page.
//
// The input is a file you keep locally and never commit; `private/` is in
// .gitignore. It looks like this:
//
//   {
//     "pages": [{
//       "title": "…",              // only shown if there are several pages
//       "field": "ripple",         // vortex | wave | scan | ripple | omit
//       "blocks": [
//         { "type": "eyebrow", "text": "…" },
//         { "type": "heading", "text": "…", "weight": "mass" },
//         { "type": "text",    "text": "…" },
//         { "type": "rule" },
//         { "type": "spec",    "entries": [["label", "value"]] },
//         { "type": "image",   "file": "photo.jpg" },
//         { "type": "gallery", "files": ["a.jpg", "b.jpg"] },
//         { "type": "gap",     "size": 48 }
//       ]
//     }]
//   }
//
// Image paths are relative to the source file and get read in and encrypted
// with everything else.
//
// The output is one AES-256-GCM blob per pair, sharing a single PBKDF2 salt,
// which the build copies to the site root and the passcode page fetches. Node's
// own crypto module does all of it, so there is nothing to install.
//
// One salt for the file rather than one per compartment, so the gate derives a
// key once however many pages there are. Two passcodes over one salt derive two
// unrelated keys, and the salt still does its real job of making this file's
// work useless against any other.
//
// What this protects, and what it does not: the blob is served to anyone who
// asks for it, so the only thing standing between a reader and the contents is
// the passcode. The iteration count below makes each guess cost real time, but
// a six-digit code is a million guesses and dedicated hardware will get there.
// Keep nothing here that would genuinely hurt to lose, and prefer a longer
// code to a shorter one — the page reads its length from one constant.

const crypto = require('crypto');
const fs = require('fs');
const path = require('path');

// OWASP's floor for PBKDF2-HMAC-SHA256. Native in both Node and the browser,
// so unlocking stays well under a second for the person who knows the code.
const ITERATIONS = 600000;

const args = process.argv.slice(2);
if (args.length === 0 || args.length % 2 !== 0) {
  console.error(
    'usage: node tool/vault.js <pages.json> <passcode> [<pages.json> <passcode> ...]',
  );
  process.exit(1);
}

const pairs = [];
for (let i = 0; i < args.length; i += 2) {
  pairs.push({ input: args[i], passcode: args[i + 1] });
}

// Two documents behind one passcode would mean the gate could open either and
// no way to say which, so the codes have to be distinct. Catching it here
// beats finding out when the wrong page opens.
const codes = new Set(pairs.map((p) => p.passcode));
if (codes.size !== pairs.length) {
  console.error('two compartments share a passcode; each one needs its own');
  process.exit(1);
}

// One salt for the whole file: the gate derives its key once and then tries
// each compartment, so adding a page costs the person entering a code nothing.
const salt = crypto.randomBytes(16);

const kb = (n) => `${(n / 1024).toFixed(0)} KB`;

const seal = ({ input, passcode }) => {
  const source = JSON.parse(fs.readFileSync(input, 'utf8'));

  // Files are named by path in the source and carried as base64 in the blob,
  // so they are encrypted alongside the words. A file in web/ would be served
  // to anyone who guessed its name, which is the one thing this is for.
  let embedded = 0;
  const inline = (file) => {
    const bytes = fs.readFileSync(path.resolve(path.dirname(input), file));
    embedded += bytes.length;
    return bytes.toString('base64');
  };

  // Anywhere in the document: { "file": "x.png" } becomes { "data": "<b64>" },
  // and { "files": [...] } becomes { "images": [...] }.
  const walk = (node) => {
    if (Array.isArray(node)) return node.forEach(walk);
    if (!node || typeof node !== 'object') return;
    if (typeof node.file === 'string') {
      node.data = inline(node.file);
      delete node.file;
    }
    if (Array.isArray(node.files)) {
      node.images = node.files.map(inline);
      delete node.files;
    }
    Object.values(node).forEach(walk);
  };
  walk(source);

  const iv = crypto.randomBytes(12);
  const key = crypto.pbkdf2Sync(passcode, salt, ITERATIONS, 32, 'sha256');

  const cipher = crypto.createCipheriv('aes-256-gcm', key, iv);
  const body = Buffer.concat([
    cipher.update(JSON.stringify(source), 'utf8'),
    cipher.final(),
  ]);

  // Web Crypto expects the tag appended to the ciphertext; Node keeps it apart.
  const data = Buffer.concat([body, cipher.getAuthTag()]);

  return { iv, data, embedded, pages: source.pages?.length ?? 0 };
};

const compartments = pairs.map(seal);

const out = path.join(__dirname, '..', 'web', 'vault.json');
fs.writeFileSync(
  out,
  JSON.stringify({
    v: 2,
    iterations: ITERATIONS,
    salt: salt.toString('base64'),
    vaults: compartments.map((c) => ({
      iv: c.iv.toString('base64'),
      data: c.data.toString('base64'),
    })),
  }),
);

const total = compartments.reduce((sum, c) => sum + c.data.length, 0);
console.log(`${out}  ${compartments.length} compartment(s)  ${kb(total)}`);
compartments.forEach((c, i) => {
  console.log(
    `  ${i + 1}. ${pairs[i].input}  ${kb(c.data.length)}` +
      (c.embedded ? `  (${kb(c.embedded)} embedded)` : ''),
  );
});
